### ISSUE TITLE
daemon with lots of z transactions in larger wallet.dat get out of sync and resyncs extremely slow

### ISSUE THREAD

**Issue by ocminer** [zcash/zcash#1749]
I have a wallet with lots of dust payments from a z_addr to lots of t_addr's 

The wallet.dat grew up to a size of about 150 MB where the daemon used up about 4 GB RAM and was constantly stuck at 100% CPU.

New transactions from the wallet didn't seem to make it to the mainchain anymore because the wallet was "behind" the mainchain a few blocks and it eventually caught up but still the payments didn't get processed.

Trying to get the wallet.dat synched on a separate machine with a fresh synched blockchain results in lots of 

- Top (height 2613): 55ba4f4527b4c4149301e2dd62a3d65eac16dd0d794f2aa5865c9fde666e056e
- New (height 1750): a8424824064141dcd5634628fd1a393041c6afc83db9642262e9d4dbbd112ebe
2016-11-01 10:24:34 Inconsistent witness cache state found for JSOutPoint(..., 1, 0)
- Cache size: 100
- Top (height 2613): 55ba4f4527b4c4149301e2dd62a3d65eac16dd0d794f2aa5865c9fde666e056e
- New (height 1750): 5334f43e5d86b00cd91c291400638d05e5eb3e0d2cdef93eb9b6df5b117b2e23
2016-11-01 10:24:42 AddToWallet ...  
2016-11-01 10:24:42 Inconsistent witness cache state found for JSOutPoint(..., 0, 0)
- Cache size: 100
- Top (height 2613): 55ba4f4527b4c4149301e2dd62a3d65eac16dd0d794f2aa5865c9fde666e056e
- New (height 1751): 3fdddd0de2cff73183f27581d22515036648237a62e7be421a36c7a0dbcde11f
2016-11-01 10:24:51 AddToWallet ...  
2016-11-01 10:24:51 Inconsistent witness cache state found for JSOutPoint(.., 0, 0)

But as you see it processes on 2-3 blocks per minute even though is not the slowest machine it's running on.

In bitcoin times you'd just dump the privkey and import it into a fresh wallet dat and that fixes this, but this doesn't seem to be possible for z_addr - is there any way of speeding up ?

Whats the future development on this ? Is this a base problem with the wallet's function, so would I have to swap wallets every 2 days on a huge pool because of this issue ?

**Comment by radix42** [comment]
Some thoughts on processing large wallets, as used by mining pool operators, from slack:

due to the comparatively huge amount of ram used to process zaddr transactions, and the larger cpu overhead, compared to bitcoin, you may need to look into running a lot more fullnodes/wallet instances for your pool.....even after any bugs are fixed 

something like take the mined coinbase UTXO taddrs and send each one to a separate worker to shove into a zaddr
I imagine various processing queues acting on the mined coinbase

**Comment by ocminer** [comment]
Actually I'm already splitting onto several "workers" or wallets for payments, however if one fails and you have 33% of the funds there, it's a big problem for pool ops (and probably also exchanges) if those cannot be regained very (!) fast and a resync takes hours currently yet we're only on block 2700ish ...

**Comment by str4d** [comment]
The issue with rescan speed is that in order to find transactions in the blockchain, every JoinSplit in every transaction in every block needs to be test-decrypted with every z-key in the wallet. So the time to do a rescan depends on:

- The number of JoinSplits in the current blockchain
- The number of z-addrs in the current wallet.

One optimisation we can make is to only test on rescan the keys that have been imported into the wallet since the last rescan. This should reduce the load, but only along the second axis. To reduce the number of blocks to be scanned, `z_importkey` would need additional information from the user about where to scan from.

> New transactions from the wallet didn't seem to make it to the mainchain anymore because the wallet was "behind" the mainchain a few blocks and it eventually caught up but still the payments didn't get processed.

This could be related to #1680, and may be improved by #1614.

**Comment by etherchain-org** [comment]
I can only second the very slow import speeds of z addresses that have lots of transactions. A z address with ~3000 tx already takes several hours to import. During the import the daemon is mostly idle at ~30-40% of CPU usage and 100MB of RAM.

Also the wallet.dat file is growing rapidly in size (already at 140MB after just a week of usage).

What is the current recommendation on how to handle use cases where one needs to send lots of transactions from a single z address (e.g. pool & exchange withdrawals)?

**Comment by ocminer** [comment]
I can second and triple that.. I've got to change wallet.dat's in a matter of 6h now to get the payments working reliably on the pool and still the nodes crash with corruptions of the wallet randomly.

A re-import on a 64 core Amazon AWS instance takes about 4 hours until it rescanned until the current block height as of now which is then the delay it takes for the users to get their coins - it's an impossible situation. 

IMHO the developers don't seem to really be interested in that issue though they have the 20% of each block, in slack I was told this is an open source project and everyone can contribute to it. 

There are thousands of senseless test cases written for every case no one would ever have thought of but the simplest things of the coin - sending and receiving funds is not working from start or at least not reliably (not to mention the disfunct z_addr -> z_addr payments).

**Comment by ebfull** [comment]
We're very interested in this problem, but there _are_ more high impact bugs we want to fix that are more common for users. There are known performance shortcomings that affect users with a large number of transactions like yourself. I personally have been travelling and won't be able to dedicate my time to these problems until next week.

Why can you not temporarily use a new z address?

**Comment by etherchain-org** [comment]
I am very disappointed by the way this issue is addressed by the Zcash team. Payments from mining pools are regularly interrupted because of it and such an issue affect several thousands of Zcash users at the same time.

I really hope you will re-prioritize the issue or at least remove the requirement that all mined funds need to go trough a z address as long as the performance issues and bugs around the z addresses are not resolved.

**Comment by nanopool** [comment]
You should re-prioritize the issue. Do you really think there are a lot of users who don't use a pool-mining or exchange services?
P.S. Got a pool wallets crashed today because of this issue.

**Comment by ebfull** [comment]
If you're a pool, there is no fundamental reason why you should be running into this performance problem.

There are two performance issues that I'm aware of:

1. Having more zaddrs makes rescan much slower.
2. Having more notes make your wallet bigger or increases the risk that it will crash for some reason. We are talking about BDB here. :)

As a pool, you should reduce the number of notes and zaddrs you have. Combine all your block rewards into a zaddr at once before you perform your batch payout.

As for prioritization, our team has not made a decision, though I have long considered this issue a problem. There are other more important bugs to fix right now. I will personally turn my attention to it soon, within days!

**Comment by pipcoinpip** [comment]
How do I get zecs with -gen=1 I have run a  node now for 48hrs nothing comes in is there a setting I am missing ?

**Comment by etherchain-org** [comment]
This is exactly war we are currently doing. All coinbase transactions are combined into one z address from which we execute the batch payments. Yet importing this single z address into a fresh wallet takes several hours to complete (approx 2.5 hours for each 1000 books). So if the node crashed or is restarted for whatever reason the wallet will be offline for at least 2.5 hours (as it seems to always scan the last 1000 blocks) which is a big issue.

**Comment by ocminer** [comment]
This is issue was fixed for me by moving the wallet(s) on a more powerful server (>128 gb ram, dual xeon)

**Comment by ebfull** [comment]
Let me clarify: using a single address is not good enough for your use case, you also need to have fewer notes. That means combining all of your block rewards at the same time, in a single transaction. If you send block rewards to a zaddr every time you receive a block, your wallet will need to maintain many merkle tree witnesses and keep them up to date as the chain advances, which is expensive right now.

**Comment by nathan-at-least** [comment]
Proposal: implement a wallet heuristic that sends a merge transaction whenever an incoming note is detected for a given address. The eventual steady state is that each address will have a single large note.

Rationale: If maintaining note witnesses is an onerous cost, this reduces the number of note witnesses towards 1-per-zaddr in the steady state case.

Pros:

- Fewer note witnesses to track and update.
- Fewer input notes to search for during payments.

Cons:
- Introduces a new "passive transaction send" behavior where txns are created and submitted without an explicit user request. This may surprise them in terms of background resource usage or prover-queue responsiveness.
- Introduces more txns than would be necessary otherwise for some cases (eg: a zaddr receives K inputs of 1 ⓩ each and then uses `z_sendmany` to send 1-fee ⓩ to K destinations would ideally require a single txn w/ K/2 JoinSplits. W/ this heuristic it would require K txns with a JS up front, plus a txn w/ ~K JS.)
- More txns on average means more funds spent on fees.

One improvement might be to let users choose between the existing note management and this heuristic, although that would make the implementation more complex than each alone, since it needs to be able to switch at runtime. Another option is to select a single heuristic for each z-addr and not allow changing heuristics in a single wallet for a given z-addr, but that introduces a UX burden.

**Comment by radix42** [comment]
Perhaps a build-time option to select which note management strategy to use? Either through ./configure or by un-commenting something in the source, similarly to how zcashd can be built to mine to a single taddr

**Comment by nathan-at-least** [comment]
radix42: That is another option. It's important however to anticipate that the same datadir may be manipulated by multiple wallet strategies, so we would need to design both to be aware of the other. I believe on benefit of that "greedily merge notes" heuristic is that it wouldn't affect any witness updates, caching, transaction scanning, etc…, so I believe it could be easily written so that switching heuristics is possible.

**Comment by ebfull** [comment]
@str4d points out that we currently increment witnesses even for spent notes, which isn't necessary. I'm going to go out on a limb and say that's most of the problem here.

**Comment by ocminer** [comment]
This is very likely looking at cpu and ram usage of the wallet. The bigger wallet.dat gets, the more ram (lots of!!)  it uses. We're talking about gigabytes of ram

**Comment by nathan-at-least** [comment]
@ebfull, @str4d: Is there any *safe* diagnostic we could ask @ocminer, @etherchain-org, and/or others to do to determine how much of the witness tracking is against spent notes?

I guess the safe way to do any diagnostic is ensure the datadir is in a consistent state, then `cp -a` it to a new temporary diagnostic location.

**Comment by nathan-at-least** [comment]
BTW- in terms of prioritization, bugs / poor-performance / UX problems which affect our users are our main priority in the short term! Sorry for the confusion.

**Comment by str4d** [comment]
@natha

**Comment by str4d** [comment]
@nathan-at-least the following patch should give us enough insight:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index e549e54..665997c 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -640,6 +640,9 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
 {
     {
         LOCK(cs_wallet);
+        size_t numNotes = 0;
+        size_t numWitnessed = 0;
+        size_t numSpent = 0;
         for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
             for (mapNoteData_t::value_type& item : wtxItem.second.mapNoteData) {
                 CNoteData* nd = &(item.second);
@@ -659,12 +662,27 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
                         nd->witnesses.pop_back();
                     }
                 }
+                // Collect stats
+                numNotes += 1;
+                if (nd->witnesses.size() > 0) {
+                    numWitnessed += 1;
+                }
+                if (IsSpent(*(nd->nullifier))) {
+                    numSpent += 1;
+                }
             }
         }
         if (nWitnessCacheSize < WITNESS_CACHE_SIZE) {
             nWitnessCacheSize += 1;
         }
 
+        // Print stats
+        LogPrint("witness", "Incrementing witnesses for block %d\n", pindex->nHeight);
+        LogPrint("witness", "- %d transactions\n", mapWallet.size());
+        LogPrint("witness", "- %d notes\n", numNotes);
+        LogPrint("witness", "  - %d previously witnessed\n", numWitnessed);
+        LogPrint("witness", "  - %d spent\n", numSpent);
+
         const CBlock* pblock {pblockIn};
         CBlock block;
         if (!pblock) {
```

**Comment by bitcartel** [comment]
@etherchain-org @ocminer Until note tracking and scanning is optimized, a practical step might be to adjust the pool payment schedule.  Looking at the data over at http://zcash.flypool.org/stats, I can see that Flypool found 149 blocks in the past 24hrs.  There are 5195 active miners using transparent addresses.  Let's assume they were the ones who worked during the past 24hrs.  With a single transaction (containing one JoinSplit), you can sweep up all 149 coinbase utxos into a single zaddr which FlyPool controls. Next, given the transaction size limit of 100,000 bytes, it will take two more transactions to payout to the 5195 miners (say ~2600 txouts in each tx).  So in this example, three transactions might have been sufficient.

**Comment by etherchain-org** [comment]
@nathan-at-least thanks for your confirmation and understanding.

Thanks a lot for your suggestions and insights into the issue. We can certainly reduce the amount of tx we use. We already defined that each tx will pay at least 50 miners. Also we can increase the interval in which we sweep coinbase funds to z addresses. Unfortunately paying out only once per day is not feasible as our main focus is to keep the pool balance low and pay out miners as soon as possible. Given the large volatility in the Zcash price also miners are interested in receiving these funds as soon as possible.

**Comment by nathan-at-least** [comment]
@str4d: Thanks for the patch!

@ocminer, et al: if you build that patch I strongly recommend this procedure which is intended to minimize production downtime, protect production wallets, and save a "pristine" diagnostic backup. The main idea is to keep this modified code away from your production wallet.

#. shutdown your production wallet's node.
#. cp -a the datadir to a new location.
#. restart the production wallet's node.

If that's too much downtime (because restarting is onerous), then you could try just a 'live' `cp -a ~/.zcash ~/DIAGNOSTIC-zcash` and we can figure out any db consistency issues after the fact.

Then:

#. *copy the backup datadir to a separate machine* off of production systems.
#. build a patched zcash on the separate diagnostic machine with the patch above.
#. make *yet another* copy of the diagnostic dir to a "live diagnostic datadir": eg: `cp -a ~/DIAGNOSTIC-SNAPSHOT-zcash ~/live-diagnostic-zcash`
#. run the modified node on the "live" diagnostic data dir, eg: `~/my-modified-zcash/src/zcashd -datadir=$HOME/live-diagnostic-zcash`.

This way if `zcashd` modified or corrupts `~/.live-diagnostic-zcash` you'll be able to start fresh with another copy of `~/DIAGNOSTIC-SNAPSHOT-zcash`, and also since it's a separate machine if, for any reason (such as running the wrong commandline) zcash modifies `~/.zcash` it won't affect your production wallet.

Doing this would help us get these statistics so we can verify @ebfull's hunch. If you don't want to do this, that's fine and we can try our own simulations. Another alternative is you can `cp -a` your datadir and then ship it to us and we can analyze it.

**Comment by etherchain-org** [comment]
@nathan-at-least I have a backup node set up where I am able to perform the diagnostics without interrupting my production system.

I applied the patch and restarted the node. You can find the log in the following gist: https://gist.github.com/etherchain-org/b08627dc97945e9ece1d3947943e0a87

As the log does not contain any timestamps, the node takes approx 10 seconds for each block to rescan.

I hope that helps.

**Comment by etherchain-org** [comment]
Any progress on that issue?

**Comment by nathan-at-least** [comment]
@etherchain-org we had a discussion about potential improvements to this performance issue yesterday. We'll look into this soon after the 1.0.3 release.

**Comment by str4d** [comment]
There are two places during rescanning that I'd expect to take time:

- Attempting to decrypt JoinSplits with z-addrs in the wallet
- Incrementing note witnesses.

Here are the first and last logged blocks from @etherchain-org's output:
```
Incrementing witnesses for block 7185
- 2635 transactions
- 2641 notes
  - 2640 previously witnessed
  - 2259 spent
[snip]
Incrementing witnesses for block 7286
- 2690 transactions
- 2696 notes
  - 2696 previously witnessed
  - 2312 spent
```

So over 85% of the notes in the wallet are spent, and still being incremented. Not incrementing those should give a decent speed-up, assuming that witness incrementing is the dominant factor. I'll do some local benchmarking to see.

**Comment by str4d** [comment]
Here's the last few runs on my laptop of a new benchmark I wrote that trial-decrypts a single JS (containing 2 dummy notes) with 10 random z-addrs:
```
    {
        "runningtime" : 0.00110200
    },
    {
        "runningtime" : 0.00110100
    },
    {
        "runningtime" : 0.00116900
    },
    {
        "runningtime" : 0.00111600
    },
    {
        "runningtime" : 0.00131400
    },
    {
        "runningtime" : 0.00113600
    },
    {
        "runningtime" : 0.00112100
    },
    {
        "runningtime" : 0.00110100
    }
]
```

So that's about 56μs per note per key. Given that in https://github.com/zcash/zcash/issues/1822#issuecomment-259194874 @etherchain-org said they only had one z-addr in their wallet, I'm certainly suspecting that witness incrementing is the dominating factor.

**Comment by str4d** [comment]
Here's the last few runs on my laptop of another benchmark that runs `IncrementNoteWitness` on 2700 wallet transactions (all with witnesses to increment):

```
    {
        "runningtime" : 0.00286800
    },
    {
        "runningtime" : 0.00279300
    },
    {
        "runningtime" : 0.00340800
    },
    {
        "runningtime" : 0.00307500
    },
    {
        "runningtime" : 0.00274500
    },
    {
        "runningtime" : 0.00278600
    },
    {
        "runningtime" : 0.00341500
    },
    {
        "runningtime" : 0.00260200
    }
]
```

**Comment by bitcartel** [comment]
Incrementing note witness might involve ReadBlockFromDisk which in turn triggers equihash verification.  Perhaps this takes up most of the time in a real-world -reindex?  Perhaps run a profiler during a full reindex (could be noisy) or benchmark around code blocks within increment note witness function (and log in csv format so you can grep and import into a spreadsheet).

**Comment by str4d** [comment]
Here's a diff that times the three components of a rescan:

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 51ed76f..27afcdf 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -642,6 +642,10 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
 {
     {
         LOCK(cs_wallet);
+        struct timeval tv_start;
+        struct timeval tv_end;
+        double elapsed;
+        gettimeofday(&tv_start, 0);
         for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
             for (mapNoteData_t::value_type& item : wtxItem.second.mapNoteData) {
                 CNoteData* nd = &(item.second);
@@ -673,6 +677,11 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
             ReadBlockFromDisk(block, pindex);
             pblock = &block;
         }
+        gettimeofday(&tv_end, 0);
+        elapsed = double(tv_end.tv_sec-tv_start.tv_sec) +
+            (tv_end.tv_usec-tv_start.tv_usec)/double(1000000);
+        LogPrintf("  - Prepared to increment witnesses in %d\n", elapsed);
+        gettimeofday(&tv_start, 0);
 
         for (const CTransaction& tx : pblock->vtx) {
             auto hash = tx.GetHash();
@@ -728,6 +737,11 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
                 }
             }
         }
+        gettimeofday(&tv_end, 0);
+        elapsed = double(tv_end.tv_sec-tv_start.tv_sec) +
+            (tv_end.tv_usec-tv_start.tv_usec)/double(1000000);
+        LogPrintf("  - Incremented witnesses in %d\n", elapsed);
+        gettimeofday(&tv_start, 0);
 
         // Update witness heights
         for (std::pair<const uint256, CWalletTx>& wtxItem : mapWallet) {
@@ -740,11 +754,20 @@ void CWallet::IncrementNoteWitnesses(const CBlockIndex* pindex,
                 assert(nWitnessCacheSize >= nd->witnesses.size());
             }
         }
+        gettimeofday(&tv_end, 0);
+        elapsed = double(tv_end.tv_sec-tv_start.tv_sec) +
+            (tv_end.tv_usec-tv_start.tv_usec)/double(1000000);
+        LogPrintf("  - Updated witness heights in %d\n", elapsed);
+        gettimeofday(&tv_start, 0);
 
         if (fFileBacked) {
             CWalletDB walletdb(strWalletFile);
             WriteWitnessCache(walletdb);
         }
+        gettimeofday(&tv_end, 0);
+        elapsed = double(tv_end.tv_sec-tv_start.tv_sec) +
+            (tv_end.tv_usec-tv_start.tv_usec)/double(1000000);
+        LogPrintf("  - Wrote witness cache in %d\n", elapsed);
     }
 }
 
@@ -1746,18 +1769,34 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
         ShowProgress(_("Rescanning..."), 0); // show rescan progress in GUI as dialog or on splashscreen, if -rescan on startup
         double dProgressStart = Checkpoints::GuessVerificationProgress(chainParams.Checkpoints(), pindex, false);
         double dProgressTip = Checkpoints::GuessVerificationProgress(chainParams.Checkpoints(), chainActive.Tip(), false);
+        struct timeval tv_start;
+        struct timeval tv_step;
+        struct timeval tv_end;
+        double elapsed;
+        gettimeofday(&tv_start, 0);
         while (pindex)
         {
             if (pindex->nHeight % 100 == 0 && dProgressTip - dProgressStart > 0.0)
                 ShowProgress(_("Rescanning..."), std::max(1, std::min(99, (int)((Checkpoints::GuessVerificationProgress(chainParams.Checkpoints(), pindex, false) - dProgressStart) / (dProgressTip - dProgressStart) * 100))));
 
             CBlock block;
+            gettimeofday(&tv_step, 0);
             ReadBlockFromDisk(block, pindex);
+            gettimeofday(&tv_end, 0);
+            elapsed = double(tv_end.tv_sec-tv_step.tv_sec) +
+                (tv_end.tv_usec-tv_step.tv_usec)/double(1000000);
+            LogPrintf("Block %d loaded in %ds\n", pindex->nHeight, elapsed);
+            gettimeofday(&tv_step, 0);
             BOOST_FOREACH(CTransaction& tx, block.vtx)
             {
                 if (AddToWalletIfInvolvingMe(tx, &block, fUpdate))
                     ret++;
             }
+            gettimeofday(&tv_end, 0);
+            elapsed = double(tv_end.tv_sec-tv_step.tv_sec) +
+                (tv_end.tv_usec-tv_step.tv_usec)/double(1000000);
+            LogPrintf("- Tried adding %d transactions to wallet in %ds\n", block.vtx.size(), elapsed);
+            gettimeofday(&tv_step, 0);
 
             ZCIncrementalMerkleTree tree;
             // This should never fail: we should always be able to get the tree
@@ -1765,6 +1804,10 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
             assert(pcoinsTip->GetAnchorAt(pindex->hashAnchor, tree));
             // Increment note witness caches
             IncrementNoteWitnesses(pindex, &block, tree);
+            gettimeofday(&tv_end, 0);
+            elapsed = double(tv_end.tv_sec-tv_step.tv_sec) +
+                (tv_end.tv_usec-tv_step.tv_usec)/double(1000000);
+            LogPrintf("- Total time to increment: %ds\n", elapsed);
 
             pindex = chainActive.Next(pindex);
             if (GetTime() >= nNow + 60) {
@@ -1772,6 +1815,11 @@ int CWallet::ScanForWalletTransactions(CBlockIndex* pindexStart, bool fUpdate)
                 LogPrintf("Still rescanning. At block %d. Progress=%f\n", pindex->nHeight, Checkpoints::GuessVerificationProgress(chainParams.Checkpoints(), pindex));
             }
         }
+        gettimeofday(&tv_end, 0);
+        elapsed = double(tv_end.tv_sec-tv_start.tv_sec) +
+            (tv_end.tv_usec-tv_start.tv_usec)/double(1000000);
+        LogPrintf("Rescan took %ds\n", elapsed);
+
         ShowProgress(_("Rescanning..."), 100); // hide progress dialog in GUI
     }
     return ret;
```

And here it is running on my laptop, with 1 z-address and no wallet transactions to increment:

```
init message: Rescanning...
Rescanning last 12 blocks (from block 12232)...
Block 12232 loaded in 0.002055s
- Tried adding 13 transactions to wallet in 0.000986s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 9e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.005091
- Total time to increment: 0.005198s
Block 12233 loaded in 0.00224s
- Tried adding 12 transactions to wallet in 0.001184s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 9e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.00512
- Total time to increment: 0.005205s
Block 12234 loaded in 0.001677s
- Tried adding 2 transactions to wallet in 0.000178s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 2e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.005125
- Total time to increment: 0.005213s
Block 12235 loaded in 0.001548s
- Tried adding 3 transactions to wallet in 0.000152s
  - Prepared to increment witnesses in 1e-06
  - Incremented witnesses in 2e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.005112
- Total time to increment: 0.005199s
Block 12236 loaded in 0.001657s
- Tried adding 3 transactions to wallet in 0.00018s
  - Prepared to increment witnesses in 1e-06
  - Incremented witnesses in 2e-06
  - Updated witness heights in 1e-06
  - Wrote witness cache in 0.005171
- Total time to increment: 0.005281s
Block 12237 loaded in 0.001943s
- Tried adding 3 transactions to wallet in 0.000498s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 5e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.005532
- Total time to increment: 0.005631s
Block 12238 loaded in 0.001819s
- Tried adding 5 transactions to wallet in 0.000454s
  - Prepared to increment witnesses in 1e-06
  - Incremented witnesses in 4e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.005255
- Total time to increment: 0.005356s
Block 12239 loaded in 0.00145s
- Tried adding 1 transactions to wallet in 1.2e-05s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 0
  - Updated witness heights in 0
  - Wrote witness cache in 0.005122
- Total time to increment: 0.005221s
Block 12240 loaded in 0.00142s
- Tried adding 2 transactions to wallet in 0.000153s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 3e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.004893
- Total time to increment: 0.005019s
Block 12241 loaded in 0.009128s
- Tried adding 3 transactions to wallet in 0.000313s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 3e-06
  - Updated witness heights in 0
  - Wrote witness cache in 0.015669
- Total time to increment: 0.015765s
Block 12242 loaded in 0.001392s
- Tried adding 1 transactions to wallet in 1.1e-05s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 0
  - Updated witness heights in 1e-06
  - Wrote witness cache in 0.004774
- Total time to increment: 0.00484s
Block 12243 loaded in 0.001356s
- Tried adding 1 transactions to wallet in 9e-06s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 0
  - Updated witness heights in 0
  - Wrote witness cache in 0.010377
- Total time to increment: 0.010416s
Block 12244 loaded in 0.002005s
- Tried adding 12 transactions to wallet in 0.001098s
  - Prepared to increment witnesses in 0
  - Incremented witnesses in 1.2e-05
  - Updated witness heights in 0
  - Wrote witness cache in 0.00507
- Total time to increment: 0.005126s
Rescan took 0.119143s
 rescan                  119ms
```

Several observations:
- Reading blocks from disk (done via `ReadBlockFromDisk()`) is significant at this scale, but clearly isn't causing the ~10s/block rescan time that @etherchain-org was seeing.
- You can clearly see the slow-down associated with attempting to decrypt JoinSplits, which will scale linearly with the number of JoinSplits in the transaction and the number of z-addrs in the wallet.
  - The only way I can currently envisage this scaling up from my wallet to one with many transactions but only one z-addr is around the insertion of transactions into the wallet hashmap, but that's unchanged from Bitcoin 0.11.2...
- The majority of time incrementing note witnesses for me is spent writing out to disk. I'll be interested to see how that scales.

@etherchain-org, if you are able to run the above patch and can capture the output from a rescan, it would help to confirm my suspicions.

**Comment by daira** [comment]
So the lowest-hanging optimisation fruit is to reduce the frequency of writing e cache to disk, right?

**Comment by str4d** [comment]
Correct; the optimisation would be to not flush the wallet in that code. But a) we'd need to ensure we could recover correctly from a crash, and b) I don't yet know how significant that actually is for a large wallet.

**Comment by etherchain-org** [comment]
@str4d Please find my log below. Most of the time seems to be spent writing the witness cache:

> init message: Rescanning...
> Rescanning last 6 blocks (from block 12391)...
> Block 12391 loaded in 0.006461s
> - Tried adding 23 transactions to wallet in 0.001149s
>   - Prepared to increment witnesses in 0.000356
>   - Incremented witnesses in 0.000196
>   - Updated witness heights in 0.000118
>   - Wrote witness cache in 4.67202
> - Total time to increment: 4.67277s
> Block 12392 loaded in 0.009681s
> - Tried adding 25 transactions to wallet in 0.001488s
>   - Prepared to increment witnesses in 0.000359
>   - Incremented witnesses in 0.000566
>   - Updated witness heights in 0.000264
>   - Wrote witness cache in 6.04963
> - Total time to increment: 6.05093s
> Block 12393 loaded in 0.010249s
> - Tried adding 23 transactions to wallet in 0.001608s
>   - Prepared to increment witnesses in 0.000363
>   - Incremented witnesses in 1.1e-05
>   - Updated witness heights in 0.000179
>   - Wrote witness cache in 5.19168
> - Total time to increment: 5.19234s
> Block 12394 loaded in 0.003185s
> - Tried adding 3 transactions to wallet in 0.000128s
>   - Prepared to increment witnesses in 0.000395
>   - Incremented witnesses in 3e-06
>   - Updated witness heights in 0.000298
>   - Wrote witness cache in 5.12844
> - Total time to increment: 5.1292s
> Block 12395 loaded in 0.010203s
> - Tried adding 24 transactions to wallet in 0.001348s
>   - Prepared to increment witnesses in 0.000364
>   - Incremented witnesses in 1e-05
>   - Updated witness heights in 0.000125
>   - Wrote witness cache in 4.97979
> - Total time to increment: 4.98035s
> Block 12396 loaded in 0.010649s
> AddToWallet 7ef0404939a332ace7f31f92a9e26dded54439266dd5606c6a0846bd5308fc03
> AddToWallet c6e34fee42d7c36608275d35e4362432632a4f880f29aa304e766c342eb2e66f
> AddToWallet 37937e7e6cf5fbff1352044e618c92a9c14af829e7f1bd0b6a9e7d42c56bcd82
> - Tried adding 25 transactions to wallet in 0.003048s
>   - Prepared to increment witnesses in 0.000372
> Inconsistent witness cache state found for JSOutPoint(c6e34fee42, 0, 0)
> - Cache size: 2
> - Top (height 12397): e9881bf45e5edb226142630bc9aaa431d187a3dd0f5517f4beb6b3b8a8ba0b8d
> - New (height 12396): 16d43ae12e491cb254918f93feeec5ad3791ce0a8f298a9b43f27d77040982de
> Inconsistent witness cache state found for JSOutPoint(37937e7e6c, 0, 0)
> - Cache size: 2
> - Top (height 12397): e9881bf45e5edb226142630bc9aaa431d187a3dd0f5517f4beb6b3b8a8ba0b8d
> - New (height 12396): a3b5583b8b8579c9f1a53cedf64dca0ddb486984e109c93bedf086a7365ecf83
>   - Incremented witnesses in 0.001672
>   - Updated witness heights in 0.000207
>   - Wrote witness cache in 5.16753
> - Total time to increment: 5.16985s
> Block 12397 loaded in 0.010173s
> AddToWallet f939ced8781c6383e260e29ce927d19690338bddeb51d3a32530df5f92471391
> - Tried adding 24 transactions to wallet in 0.001266s
>   - Prepared to increment witnesses in 0.000358
>   - Incremented witnesses in 1.1e-05
>   - Updated witness heights in 0.000173
>   - Wrote witness cache in 5.20563
> - Total time to increment: 5.20627s
>

**Comment by str4d** [comment]
@etherchain-org fantastic, thanks! This tells me two things:

- The block-loading, note-finding and witness-incrementing appear to scale as expected.
- We can get huge performance gains by not flushing the wallet with each block.

**Comment by str4d** [comment]
@etherchain-org could you try #1891 on a clone of your wallet and see whether it speeds up rescanning for you?

**Comment by etherchain-org** [comment]
@str4d I tried #1891 but did not see any significant improvement. It still takes +/- 10secs to process a block.

**Comment by str4d** [comment]
@etherchain-org interesting that there was no effect on `-rescan`, because the wallet-writing should have decreased from once per block to once every 2 seconds with that PR.

Could you please try #1904 on a clone of your wallet, along with the patch in https://github.com/zcash/zcash/issues/1749#issuecomment-261426652, and report back on the effect of a `-rescan`? This PR should not write to disk at all during a rescan.

**Comment by etherchain-org** [comment]
@str4d the new PR #1904 works extremely fast. Importing a key that previously took hours is done now in just 119.195s.

Here is an excerpt of the log file:
`
Block 13300 loaded in 0.003127s
AddToWallet 7b3a950b5f58281fcd7723f900fa2669f80582ec088fa3ceead21a267954c7f9  new
- Tried adding 2 transactions to wallet in 0.004799s
  - Prepared to increment witnesses in 0.002831
  - Incremented witnesses in 0.00296
  - Updated witness heights in 0.000154
- Total time to increment: 0.005983s
Block 13301 loaded in 0.002058s
AddToWallet 444958406df468f747eb4cc44e5df0f31109592125c48a32b0f2f06dd2948205  new
- Tried adding 4 transactions to wallet in 0.004445s
  - Prepared to increment witnesses in 0.002798
  - Incremented witnesses in 0.002109
  - Updated witness heights in 0.000155
- Total time to increment: 0.005097s
Block 13302 loaded in 0.002425s
AddToWallet fa5c10dfa134d30fa38c2cd78fa9e6056e6c4c90a7897b9dbde5700afa0366bd  new
AddToWallet fb77c3b8d2921cc071366c55f16de522018ef03dfca18b61eef31787c9ae451a  new
- Tried adding 8 transactions to wallet in 0.011157s
  - Prepared to increment witnesses in 0.002757
  - Incremented witnesses in 0.020268
  - Updated witness heights in 0.000133
- Total time to increment: 0.023196s
Block 13303 loaded in 0.001697s
- Tried adding 1 transactions to wallet in 7e-06s
  - Prepared to increment witnesses in 0.00266
  - Incremented witnesses in 1e-06
  - Updated witness heights in 0.000145
- Total time to increment: 0.002836s
Block 13304 loaded in 0.001565s
AddToWallet 1a5037e7e6e293cc691465cd3fed55e5b57378c9164b8e9c787a909053eeb7db  new
- Tried adding 4 transactions to wallet in 0.003537s
  - Prepared to increment witnesses in 0.002764
  - Incremented witnesses in 0.002633
  - Updated witness heights in 0.000137
- Total time to increment: 0.005566s
Block 13305 loaded in 0.001858s
AddToWallet c8c0b5bef52f8d3c44bbab9f9a80e1e996d8a044a14272b1fb9738c8b06749c3  new
AddToWallet 94535582af906778ef8d5f0d7a4c2f153fea4453eb2d135a4a90f03cbea347ed  new
- Tried adding 6 transactions to wallet in 0.008404s
  - Prepared to increment witnesses in 0.002629
  - Incremented witnesses in 0.006005
  - Updated witness heights in 0.000137
- Total time to increment: 0.008812s
Block 13306 loaded in 0.001538s
- Tried adding 1 transactions to wallet in 9e-06s
  - Prepared to increment witnesses in 0.002528
  - Incremented witnesses in 0
  - Updated witness heights in 0.000138
- Total time to increment: 0.002694s
Block 13307 loaded in 0.002004s
AddToWallet 2879e3ffa6565fd83598ed3364a1764b513baf29ed04d1e22d14d20b096debc9  new
AddToWallet b9455535afc5abc9093389e1a970c6ace624dfdb44e48b8e7c4479e8c287bb42  new
- Tried adding 12 transactions to wallet in 0.007602s
  - Prepared to increment witnesses in 0.002644
  - Incremented witnesses in 0.008803
  - Updated witness heights in 0.000129
- Total time to increment: 0.011608s
Block 13308 loaded in 0.00182s
- Tried adding 3 transactions to wallet in 0.00077s
  - Prepared to increment witnesses in 0.002675
  - Incremented witnesses in 0.003883
  - Updated witness heights in 0.000127
- Total time to increment: 0.006716s
Block 13309 loaded in 0.001966s
- Tried adding 12 transactions to wallet in 0.001695s
  - Prepared to increment witnesses in 0.002633
  - Incremented witnesses in 0.01426
  - Updated witness heights in 0.000112
- Total time to increment: 0.017037s
Block 13310 loaded in 0.001492s
- Tried adding 5 transactions to wallet in 0.000556s
  - Prepared to increment witnesses in 0.0025
  - Incremented witnesses in 0.003608
  - Updated witness heights in 0.00011
- Total time to increment: 0.006246s
Block 13311 loaded in 0.00154s
AddToWallet 39f74fa583deee6cfc9afa90059a4d878ec6e019b978e76834c64fcbaa3df1c1  new
- Tried adding 5 transactions to wallet in 0.00425s
  - Prepared to increment witnesses in 0.002507
  - Incremented witnesses in 0.004946
  - Updated witness heights in 0.000107
- Total time to increment: 0.007633s
Block 13312 loaded in 0.001295s
- Tried adding 1 transactions to wallet in 7e-06s
  - Prepared to increment witnesses in 0.002437
  - Incremented witnesses in 1e-06
  - Updated witness heights in 0.000158
- Total time to increment: 0.00263s
Block 13313 loaded in 0.001303s
- Tried adding 1 transactions to wallet in 9e-06s
  - Prepared to increment witnesses in 0.002562
  - Incremented witnesses in 1e-06
  - Updated witness heights in 0.000182
- Total time to increment: 0.002779s
Block 13314 loaded in 0.00131s
- Tried adding 2 transactions to wallet in 1e-05s
  - Prepared to increment witnesses in 0.002527
  - Incremented witnesses in 1e-06
  - Updated witness heights in 0.000199
- Total time to increment: 0.00276s
Block 13315 loaded in 0.0016s
AddToWallet c3693f46db6afe5721a7eab67f812471038aea140879c2ffcef2dce39636b099  new
- Tried adding 9 transactions to wallet in 0.004356s
  - Prepared to increment witnesses in 0.002421
  - Incremented witnesses in 0.007858
  - Updated witness heights in 0.000123
- Total time to increment: 0.010431s
Block 13316 loaded in 0.001353s
- Tried adding 6 transactions to wallet in 0.00025s
  - Prepared to increment witnesses in 0.002408
  - Incremented witnesses in 0.001403
  - Updated witness heights in 0.000117
- Total time to increment: 0.003954s
Block 13317 loaded in 0.00158s
- Tried adding 10 transactions to wallet in 0.000779s
  - Prepared to increment witnesses in 0.002474
  - Incremented witnesses in 0.008575
  - Updated witness heights in 0.000109
- Total time to increment: 0.011185s
Block 13318 loaded in 0.001441s
AddToWallet 5b60de433355c128d823c5e32bdb02b5b2d7ba266e6d7b46ec23a51e86e12f48  new
- Tried adding 2 transactions to wallet in 0.003857s
  - Prepared to increment witnesses in 0.002511
  - Incremented witnesses in 0.003509
  - Updated witness heights in 0.000115
- Total time to increment: 0.006163s
Block 13319 loaded in 0.001263s
- Tried adding 2 transactions to wallet in 1.1e-05s
  - Prepared to increment witnesses in 0.002436
  - Incremented witnesses in 1e-06
  - Updated witness heights in 0.000119
- Total time to increment: 0.002579s
Block 13320 loaded in 0.001253s
AddToWallet a5b24dec0d1b5e6e4dd9e7b5f23488131b3675e874c040b39ba67806c20289ac  new
- Tried adding 3 transactions to wallet in 0.003581s
  - Prepared to increment witnesses in 0.002537
  - Incremented witnesses in 0.001394
  - Updated witness heights in 0.000119
- Total time to increment: 0.004076s
Block 13321 loaded in 0.001731s
`

**Comment by str4d** [comment]
@ocminer @etherchain-org after 1.0.4 is released, please report back here on whether your performance issues are improved or resolved.

**Comment by str4d** [comment]
@ocminer @etherchain-org were your performance issues improved or resolved by 1.0.4?

**Comment by ocminer** [comment]
For me it was already fixed with 1.0.3 by moving onto a much (!) faster machine (48c, 128gb ram) but 1.0.4 is even better. However with wallet files > 200 MB (lots of transactions) the wallet still gets "slowish" in responsiveness. Creating a new wallet.dat, sending all funds there and then working from that new wallet fixes the issue for a month or so (until the wallet grows to 200 MB again..)

**Comment by bitcartel** [comment]
We're not doing anything for this in 1.0.5 but we seek to make improvements in 1.0.6.
